### DISCUSSION TITLE
Component-State Styling of SVGs

### DISCUSSION THREAD

**Discussion by justinmcbride** [nativewind/nativewind#143] [Q&A] [answered]
Hi there,

I was wondering how I could combine the Component State styling (specifically looking for `component-hover`) to affect the `fill` (or really, any of the props) for the SVG classes.

This might show the code that I was trying, using [react-native-heroicons](https://www.npmjs.com/package/react-native-heroicons)):
```tsx
<Pressable className='component'>
  <PlusIcon size={32} className='component-hover:fill-red-500' />
</Pressable>
```

Referencing the [SVG documentation here](https://tailwindcss-react-native.vercel.app/tailwind/svg/fill) I can see that you need to take another approach to the SVG `fill` by using a prop instead of a normal class. I'm just not sure of the best way to proceed here. Any advice is welcome.

Thank you!

**Comment by justinmcbride** [comment, marked as answer]
I was able to figure it out pretty quickly after posing this question, using the `styled` API that you have [documented here](https://tailwindcss-react-native.vercel.app/api/styled)!

This is the code that ended up working:
```tsx
const StyledPlus = styled(PlusIcon, {classProps: ["fill"]})

const MyPlusIcon = ({ fill, ...props }) => {
  return (
    <StyledPlus {...props } fill={fill} />
  )
}
<Pressable className='component'>
  <MyPlusIcon size={32} fill='fill-white component-hover:fill-blue-500' />
</Pressable>
```

Honestly, I'm not sure that I exactly understand how the `styled` API transforms the properties and what magic happens behind the scenes, but that may just come with time.

Perhaps my only remaining question might be which type I need to declare on the following line:
```tsx
const MyPlusIcon = ({ fill, ...props }) => {
```
It produces the error `Binding element 'fill' implicitly has an 'any' type.ts(7031)` during the `tsc` command. I don't believe that this would be an issue of the `nativewind` library, but rather just a general issue when one doesn't have access to the props interface for the type that you're styling. 

In the `react-native-heroicons` case, it [looks like](https://github.com/ecklf/react-native-heroicons/blob/main/bin/template.ts) the `Props` interface isn't exported, and as such I can't extend upon it to add something akin to
```tsx
interface IconProps extends PlusIcon.Props {
  fill: string;
}
```
to properly define the type for the above erroring line.

**Reply by marklawlor** [reply]
Glad you got it working!

At a technical level `classProps` does one of two things depending on the platform. 

For web, it simply adds the prop to the className string. 

For native, a pseudo coded version looks like this

```jsx
<MyPlusIcon fill={StyleSheet.flatten(transformTailwind('fill-white component-hover:fill-blue-500'))['fill']} >
```

As for the types, you will need to use [declaration-merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html). Since the props are not exported, you can simply merge onto the `SvgProps` that it extends from

**Reply by justinmcbride** [reply]
Thank you for the response! It was very informative and I understand what's happening now. Appreciate your work on this project.

**Comment by danstepanov** [comment]
Closing this discussion, if this is still relevant, feel free to open a new one.
